hatchery/src/snapshot/module_snapshot.rs

```diff
@@ -0,0 +1,253 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+//
+// Copyright (c) DUSK NETWORK. All rights reserved.
+
+use crate::error::Error;
+use crate::storage_helpers::ByteArrayWrapper;
+use crate::Error::PersistenceError;
+use byteorder::{LittleEndian, ReadBytesExt, WriteBytesExt};
+use qbsdiff::Bsdiff;
+use qbsdiff::Bspatch;
+use rand::Rng;
+use rkyv::{Archive, Deserialize, Serialize};
+use std::fs::OpenOptions;
+use std::io::{Read, Write};
+use std::mem;
+use std::path::{Path, PathBuf};
+
+const COMPRESSION_LEVEL: i32 = 11;
+pub const MODULE_SNAPSHOT_ID_BYTES: usize = 32;
+#[derive(
+    Debug,
+    PartialEq,
+    Eq,
+    Archive,
+    Serialize,
+    Deserialize,
+    PartialOrd,
+    Ord,
+    Hash,
+    Clone,
+    Copy,
+)]
+pub struct ModuleSnapshotId([u8; MODULE_SNAPSHOT_ID_BYTES]);
+impl ModuleSnapshotId {
+    pub fn as_bytes(&self) -> &[u8] {
+        &self.0
+    }
+    pub fn random() -> ModuleSnapshotId {
+        ModuleSnapshotId(
+            rand::thread_rng().gen::<[u8; MODULE_SNAPSHOT_ID_BYTES]>(),
+        )
+    }
+}
+impl From<[u8; 32]> for ModuleSnapshotId {
+    fn from(array: [u8; 32]) -> Self {
+        ModuleSnapshotId(array)
+    }
+}
+
+pub trait ModuleSnapshotLike {
+    fn path(&self) -> &PathBuf;
+    /// Read's module snapshot's content into buffer
+    fn read(&self) -> Result<Vec<u8>, Error> {
+        let mut f = std::fs::File::open(self.path().as_path())
+            .map_err(PersistenceError)?;
+        let metadata = std::fs::metadata(self.path().as_path())
+            .map_err(PersistenceError)?;
+        let mut buffer = vec![0; metadata.len() as usize];
+        f.read(buffer.as_mut_slice()).map_err(PersistenceError)?;
+        Ok(buffer)
+    }
+}
+
+pub struct MemoryPath {
+    path: PathBuf,
+}
+
+impl MemoryPath {
+    pub fn new(path: impl AsRef<Path>) -> Self {
+        MemoryPath {
+            path: path.as_ref().to_path_buf(),
+        }
+    }
+}
+
+impl ModuleSnapshotLike for MemoryPath {
+    fn path(&self) -> &PathBuf {
+        &self.path
+    }
+}
+
+fn combine_module_snapshot_names(
+    module_name: impl AsRef<str>,
+    snapshot_name: impl AsRef<str>,
+) -> String {
+    format!("{}_{}", module_name.as_ref(), snapshot_name.as_ref())
+}
+
+fn module_snapshot_id_to_name(module_snapshot_id: ModuleSnapshotId) -> String {
+    format!("{}", ByteArrayWrapper(module_snapshot_id.as_bytes()))
+}
+
+pub struct ModuleSnapshot {
+    path: PathBuf,
+    id: ModuleSnapshotId,
+}
+
+impl ModuleSnapshot {
+    pub(crate) fn new(memory_path: &MemoryPath) -> Result<Self, Error> {
+        let module_snapshot_id: ModuleSnapshotId = ModuleSnapshotId::from(
+            *blake3::hash(memory_path.read()?.as_slice()).as_bytes(),
```

Comment:
This hashing includes the argument buffer, which should be ignored when hashing.